### PR TITLE
[IMP] populate database

### PR DESCRIPTION
**Pull request by Xavier-Do**
Some use case like testing performance or upgrade scripts required a database
with prefilled data, covering basic corner cases. A solution can be to
create data a procedural way.

This commit proposes an API to easily populate a model, usually by giving a
list of possible values for each field or by giving a compute method that will
be based on raw values of other fields.

The basic way to define how to populate a new field is to override `_populate_factories`,
a method that returns a sequence of pairs `(field_name, factory)`.

The definition of a field is a "factory", a function that returns a neverending iterator
combining its value(s) with the values of the iterator given in parameter.
Some factory helpers are given in `tools.populate.py`:
- `iterate(vals, weighs)` ensures that one record is created for each value
by iterating on them, then resumes as `random.choice` on those vals following weights
once the first iteration is finished.
- `cartesian(vals, weights)` makes a cartesian product of its own values with the values
of its input iterator, then resumes as a randomized generator.
- `compute(function)` calls the given function with the current values dict and a random object,
and assigns the current field to the returned value.
- ...

Each iterator yields dictionaries of field values, and the factory should add a
value for the current field(s).  The yielded dictionaries also contain a pseudo_field
`"__complete"`, that indicates whether this step is some randomized data
to reach the expected count of records.  A falsy value indicates that the iterator
is still covering mandatory cases.  This indicates whether a cartesian product is
finished, or an `iterate` has consumed all its values.

The order of the factories is quite important, since some computed fields may need
other fields to be defined, and `cartesian` factories should always be at the beginning
to avoid having too many combination.  That is why the factories are given as a list of
pairs instead of a dictionary; this makes it easier to insert elements at any place.

Example:
field A: cartesian([T, F])
field B: cartesian([0, 1])
field C: iterate([a, b, c, d, e])
field D: compute(1-B)

_c is shortcut for __complete
_ is a random value, or result of a random value

```
iter | root  | field A  | field B | field C     | field D   | result
0     {_c:F}  {... A:T}  {... B:0} {...C:a}      {...D:1}    T,0,a,1 complete:False
                         {... B:1} {...C:b}      {...D:0}    T,1,b,0 complete:False
              {... A:F}  {... B:0} {...C:c}      {...D:1}    F,0,c,1 complete:False
                         {... B:1} {...C:d}      {...D:0}    F,1,d,0 complete:False
1     {_c:T}  {... A:_}  {... B:_} {...C:e,_c:F} {...D:_}    _,_,e,_ complete:False
2     {_c:T}  {... A:_}  {... B:_} {...C:_}      {...D:_}    _,_,_,_ complete:True
```

### CODE REVIEW

File: odoo/tools/populate.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+def constant(val, formater=format_str):
+    """ Return a factory for an iterator of values dicts that sets the field
+    to the given value in each input dict.
+    """
+    def generate(iterator, field_name, _):
+        counter = 0
+        for values in iterator:
+            values[field_name] = formater(val, counter, values)
+            yield values
+            counter += 1
```
Comment (rco-odoo) [inline odoo/tools/populate.py]: This one can use `enumerate`, it is safe in this case.

File: odoo/tools/populate.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    """
+    def generate(iterator, field_name, model_name):
+        counter = 0
+        for val in vals: # iteratable order is important, shortest first
+            values = next(iterator)
+            values[field_name] = formater(val, counter, values)
+            values['__complete'] = False
+            yield values
+            counter += 1
+        yield from (then if then else randomize(vals, weights, seed, formater, counter))(iterator, field_name, model_name)
```
Comment (rco-odoo) [inline odoo/tools/populate.py]: Ahem, readability counts
```suggestion
        factory = then or randomize(vals, weights, seed, formater, counter)
        yield from factory(iterator, field_name, model_name)
```

File: odoo/tools/populate.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    """
+    def generate(iterator, field_name, model_name):
+        counter = 0
+        for values in iterator:
+            if values['__complete']:
+                break  # will consume and lose an element, (complete so a filling element). If it is a problem, use peekable instead.
+            for val in vals:
+                yield {**values, field_name: formater(val, counter, values)}
+            counter += 1
+        yield from (then if then else randomize(vals, weights, seed, formater, counter))(iterator, field_name, model_name)
```
Comment (rco-odoo) [inline odoo/tools/populate.py]:
```suggestion
        factory = then or randomize(vals, weights, seed, formater, counter)
        yield from factory(iterator, field_name, model_name)
```

File: odoo/tools/populate.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+def randomize(vals, weights=None, seed=False, formater=format_str, counter_offset=0):
+    """ Return a factory for an iterator of values dicts with pseudo-randomly
+    chosen values (among ``vals``) for a field.
+    """
+    def generate(iterator, field_name, model_name):
+        r = Random('%s+field+%s' % (model_name, seed or field_name))
+        for counter, values in enumerate(iterator):
+            val = r.choices(vals, weights)[0]
+            values[field_name] = formater(val, counter+counter_offset, values)
```
Comment (rco-odoo) [inline odoo/tools/populate.py]:
```suggestion
            values[field_name] = formater(val, counter + counter_offset, values)
```

File: odoo/tools/populate.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+
+def root_factory():
+    """ Return a generator with empty values dictionaries (except for the flag ``__complete``). """
+    yield {'__complete': False}
+    while True:
+        yield {'__complete': True}
+
+
+def randomize(vals, weights=None, seed=False, formater=format_str, counter_offset=0):
```
Comment (rco-odoo) [inline odoo/tools/populate.py]: The right spelling is `formatter`.

File: odoo/models.py [hunk 6]
```diff
@@ -6181,6 +6183,100 @@ def _get_placeholder_filename(self, field=None):
...
+    @property
+    def _populate_dependencies(self):
+        """ Return a list of model names to populate before this model.  The base implementation
+        automatically determines dependencies based on the comodels of the many2one fields
+        on this model.  A field's comodel is omitted if it create a cycle in dependencies.
+        This property should be overridden to add missing dependencies.
+        """
+        deps = []
+        field_black_list = ('create_uid', 'write_uid')
+        def is_electible_dependency(field):
```
Comment (rco-odoo) [inline odoo/models.py]: I suggest `is_eligible_dependency`

File: odoo/models.py [hunk 7]
```diff
@@ -33,6 +33,7 @@
 import pytz
 import re
 import uuid
+import random
```
Comment (rco-odoo) [inline odoo/models.py]: not used?

File: odoo/models.py [hunk 8]
```diff
@@ -6181,6 +6183,100 @@ def _get_placeholder_filename(self, field=None):
             complete path to access it (eg: module/path/to/image.png).
         """
         return 'placeholder.png'
+    def _populate_factories(self):
```
Comment (rco-odoo) [inline odoo/models.py]: Doin' my *black*...
```suggestion

    def _populate_factories(self):
```

File: odoo/addons/base/populate/res_user.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    _inherit = "res.users"
+
+    _populate_sizes = {"small": 10, "medium": 2000, "large": 10000}
+
+    _populate_dependencies = ["res.partner"]
+
+    def _populate_factories(self):
+        partner_ids = self.env.registry.populated_models["res.partner"]
+
+        def partner_id_callable(random=None, **kwargs):
```
Comment (rco-odoo) [inline odoo/addons/base/populate/res_user.py]:
```suggestion
        def get_partner_id(random=None, **kwargs):
```

**Comment by rco-odoo** [comment]
I edited the PR's description to make it easier to understand.
Can you reformulate the commit message after that?
Thanks.

**Comment by Xavier-Do** [comment]
robodoo r+ p=0

**Comment by Xavier-Do** [comment]
robodoo r-

**Comment by Xavier-Do** [comment]
robodoo r+ p=0

**Comment by Xavier-Do** [comment]
robodoo r-

**Comment by Xavier-Do** [comment]
robodoo r+

**Comment by Xavier-Do** [comment]
robodoo r+

**Comment by Xavier-Do** [comment]
robodoo rebase-ff

**Pull request closed without merging**